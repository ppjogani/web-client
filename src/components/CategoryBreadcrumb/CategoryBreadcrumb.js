import React from 'react';
import { array, bool, object, oneOfType, string } from 'prop-types';
import classNames from 'classnames';

import { FormattedMessage } from '../../util/reactIntl';
import { createSlug } from '../../util/urlHelpers';
import { NamedLink } from '../../components';

import css from './CategoryBreadcrumb.module.css';

/**
 * Parse category hierarchy from category string or object
 * Supports formats like:
 * - "Baby Care > Feeding > Bottles" (string with separators)
 * - ["Baby Care", "Feeding", "Bottles"] (array)
 * - { main: "Baby Care", sub: "Feeding", specific: "Bottles" } (object)
 */
const parseCategoryHierarchy = (category) => {
  if (!category) return [];

  // If it's a string with separators (>, /, |, etc.)
  if (typeof category === 'string') {
    // Split by common separators and clean whitespace
    const separators = /[>\/|â†’\-]/;
    if (separators.test(category)) {
      return category.split(separators).map(cat => cat.trim()).filter(Boolean);
    }
    // Single category
    return [category];
  }

  // If it's an array
  if (Array.isArray(category)) {
    return category.filter(Boolean);
  }

  // If it's an object with hierarchy properties
  if (typeof category === 'object') {
    const hierarchy = [];

    // Support both legacy format (main, sub, specific, detailed) and level-based format
    if (category.main) hierarchy.push(category.main);
    if (category.sub) hierarchy.push(category.sub);
    if (category.specific) hierarchy.push(category.specific);
    if (category.detailed) hierarchy.push(category.detailed);

    // Support level-based format (level1, level2, level3, etc.)
    const levelKeys = Object.keys(category)
      .filter(key => key.startsWith('level'))
      .sort(); // Sort to ensure correct order (level1, level2, level3, etc.)

    levelKeys.forEach(key => {
      if (category[key]) hierarchy.push(category[key]);
    });

    return hierarchy;
  }

  return [];
};

/**
 * Build category path for search filtering
 * Creates cumulative category paths for each level
 */
const buildCategoryPaths = (hierarchy) => {
  const paths = [];
  let currentPath = '';

  hierarchy.forEach((category, index) => {
    if (index === 0) {
      currentPath = category;
    } else {
      currentPath += ` > ${category}`;
    }

    paths.push({
      name: category,
      fullPath: currentPath,
      slug: createSlug(category),
      level: index
    });
  });

  return paths;
};

/**
 * CategoryBreadcrumb component displays hierarchical category navigation
 * with links to filter by each category level
 */
const CategoryBreadcrumb = props => {
  const {
    rootClassName,
    className,
    category,
    showHomeLink = true,
    separator = '>',
  } = props;

  const hierarchy = parseCategoryHierarchy(category);

  if (!hierarchy || hierarchy.length === 0) {
    return null;
  }

  const categoryPaths = buildCategoryPaths(hierarchy);
  const classes = classNames(rootClassName || css.root, className);

  return (
    <nav className={classes} aria-label="Category breadcrumb">
      <ol className={css.breadcrumbList}>
        {showHomeLink && (
          <li className={css.breadcrumbItem}>
            <NamedLink
              name="SearchPage"
              className={css.homeLink}
              title="View all products"
            >
              <FormattedMessage id="CategoryBreadcrumb.home" />
            </NamedLink>
            <span className={css.separator} aria-hidden="true">
              {separator}
            </span>
          </li>
        )}

        {categoryPaths.map((categoryPath, index) => {
          const isLast = index === categoryPaths.length - 1;

          return (
            <li key={categoryPath.fullPath} className={css.breadcrumbItem}>
              {isLast ? (
                <span className={css.currentCategory} aria-current="page">
                  {categoryPath.name}
                </span>
              ) : (
                <>
                  <NamedLink
                    name="SearchPage"
                    to={{
                      search: `?pub_category=${encodeURIComponent(categoryPath.fullPath)}`
                    }}
                    className={css.categoryLink}
                    title={`View all products in ${categoryPath.name}`}
                  >
                    {categoryPath.name}
                  </NamedLink>
                  <span className={css.separator} aria-hidden="true">
                    {separator}
                  </span>
                </>
              )}
            </li>
          );
        })}
      </ol>
    </nav>
  );
};

CategoryBreadcrumb.defaultProps = {
  rootClassName: null,
  className: null,
  showHomeLink: true,
  separator: '>',
};

CategoryBreadcrumb.propTypes = {
  rootClassName: string,
  className: string,
  category: oneOfType([
    string,
    array,
    object,
  ]).isRequired,
  showHomeLink: bool,
  separator: string,
};

export default CategoryBreadcrumb;