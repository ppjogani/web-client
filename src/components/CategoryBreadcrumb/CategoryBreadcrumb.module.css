@import '../../styles/customMediaQueries.css';

.root {
  margin: 16px 0;
  padding: 0 24px;

  @media (--viewportMedium) {
    margin: 24px 0;
    padding: 0;
  }
}

.breadcrumbList {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  list-style: none;
  margin: 0;
  padding: 0;
  gap: 8px;

  @media (--viewportMedium) {
    gap: 12px;
  }
}

.breadcrumbItem {
  display: flex;
  align-items: center;
  gap: 8px;

  @media (--viewportMedium) {
    gap: 12px;
  }
}

.homeLink,
.categoryLink {
  color: var(--colorGrey600);
  text-decoration: none;
  font-size: 14px;
  font-weight: var(--fontWeightMedium);
  padding: 4px 8px;
  border-radius: 4px;
  transition: all 0.2s ease;

  &:hover {
    color: var(--colorGrey800);
    background-color: var(--colorGrey100);
    text-decoration: none;
  }

  &:focus {
    outline: 2px solid var(--colorPrimary);
    outline-offset: 2px;
  }

  @media (--viewportMedium) {
    font-size: 15px;
    padding: 6px 12px;
  }
}

.homeLink {
  color: var(--colorPrimary);
  font-weight: var(--fontWeightSemiBold);

  &:hover {
    color: var(--colorPrimaryDark);
    background-color: var(--colorPrimary50);
  }
}

.currentCategory {
  color: var(--colorGrey800);
  font-size: 14px;
  font-weight: var(--fontWeightSemiBold);
  padding: 4px 8px;

  @media (--viewportMedium) {
    font-size: 15px;
    padding: 6px 12px;
  }
}

.separator {
  color: var(--colorGrey400);
  font-size: 12px;
  font-weight: var(--fontWeightMedium);
  user-select: none;

  @media (--viewportMedium) {
    font-size: 14px;
  }
}

/* Responsive behavior for mobile */
@media (--viewportSmall) {
  .breadcrumbList {
    font-size: 13px;
  }

  .homeLink,
  .categoryLink,
  .currentCategory {
    font-size: 13px;
    padding: 3px 6px;
  }

  .separator {
    font-size: 11px;
  }
}

/* Ensure proper wrapping on very small screens */
@media (max-width: 480px) {
  .breadcrumbList {
    flex-direction: column;
    align-items: flex-start;
    gap: 4px;
  }

  .breadcrumbItem {
    gap: 6px;
  }

  .separator {
    display: none;
  }

  .breadcrumbItem:not(:last-child)::after {
    content: '>';
    color: var(--colorGrey400);
    font-size: 11px;
    margin-left: 6px;
  }
}