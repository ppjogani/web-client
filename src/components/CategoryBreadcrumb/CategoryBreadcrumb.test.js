import React from 'react';
import '@testing-library/jest-dom';

import { renderWithProviders, testingLibrary } from '../../util/testHelpers';
import CategoryBreadcrumb from './CategoryBreadcrumb';

const { screen } = testingLibrary;

describe('CategoryBreadcrumb', () => {
  it('renders nothing when category is null or undefined', () => {
    // Suppress PropTypes warning for this test
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const { container } = renderWithProviders(<CategoryBreadcrumb category={null} />);
    expect(container.firstChild).toBeNull();

    console.error = originalConsoleError;
  });

  it('renders single category as string', () => {
    renderWithProviders(<CategoryBreadcrumb category="Baby Clothing" />);

    expect(screen.getByText('Baby Clothing')).toBeInTheDocument();
    expect(screen.getByRole('link')).toHaveAttribute('href', '/s');
  });

  it('parses category string with separators', () => {
    renderWithProviders(<CategoryBreadcrumb category="Baby Products > Clothing > Organic Cotton" />);

    expect(screen.getByText('Baby Products')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
    expect(screen.getByText('Organic Cotton')).toBeInTheDocument();
  });

  it('handles different separators (slash, pipe, arrow)', () => {
    renderWithProviders(<CategoryBreadcrumb category="Baby/Clothing|Organicâ†’Cotton" />);

    expect(screen.getByText('Baby')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
    expect(screen.getByText('Organic')).toBeInTheDocument();
    expect(screen.getByText('Cotton')).toBeInTheDocument();
  });

  it('renders array category format', () => {
    const categoryArray = ['Baby Products', 'Clothing', 'Organic'];
    renderWithProviders(<CategoryBreadcrumb category={categoryArray} />);

    expect(screen.getByText('Baby Products')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
    expect(screen.getByText('Organic')).toBeInTheDocument();
  });

  it('renders object category format with hierarchical structure', () => {
    const categoryObject = {
      level1: 'Baby Products',
      level2: 'Clothing',
      level3: 'Organic Cotton'
    };
    renderWithProviders(<CategoryBreadcrumb category={categoryObject} />);

    expect(screen.getByText('Baby Products')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
    expect(screen.getByText('Organic Cotton')).toBeInTheDocument();
  });

  it('creates correct search links for each category level', () => {
    renderWithProviders(<CategoryBreadcrumb category="Baby Products > Clothing > Organic" />);

    const links = screen.getAllByRole('link');

    // Should have home link and category links (excluding last item which is not a link)
    expect(links.length).toBeGreaterThan(0);

    // Check that some links exist with correct href patterns
    const linkWithCategory = screen.getByRole('link', { name: 'Baby Products' });
    expect(linkWithCategory).toHaveAttribute('href', '/s?pub_category=Baby%20Products');
  });

  it('applies custom className', () => {
    const { container } = renderWithProviders(
      <CategoryBreadcrumb category="Baby" className="custom-breadcrumb" />
    );

    expect(container.firstChild).toHaveClass('custom-breadcrumb');
  });

  it('handles empty category strings gracefully', () => {
    const { container } = renderWithProviders(<CategoryBreadcrumb category="" />);
    expect(container.firstChild).toBeNull();
  });

  it('trims whitespace from category segments', () => {
    renderWithProviders(<CategoryBreadcrumb category="  Baby Products  >  Clothing  " />);

    expect(screen.getByText('Baby Products')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
  });

  it('filters out empty segments', () => {
    renderWithProviders(<CategoryBreadcrumb category="Baby Products > > Clothing" />);

    expect(screen.getByText('Baby Products')).toBeInTheDocument();
    expect(screen.getByText('Clothing')).toBeInTheDocument();
  });

  it('builds cumulative category paths correctly', () => {
    renderWithProviders(<CategoryBreadcrumb category="Level1 > Level2 > Level3" />);

    const level1Link = screen.getByRole('link', { name: 'Level1' });
    const level2Link = screen.getByRole('link', { name: 'Level2' });

    expect(level1Link).toHaveAttribute('href', '/s?pub_category=Level1');
    expect(level2Link).toHaveAttribute('href', '/s?pub_category=Level1%20%3E%20Level2');
  });

  describe('routing and URL encoding', () => {
    it('handles special characters in category names', () => {
      renderWithProviders(<CategoryBreadcrumb category="Baby & Toddler > Clothing & Accessories" />);

      const babyLink = screen.getByRole('link', { name: 'Baby & Toddler' });
      expect(babyLink).toHaveAttribute('href', '/s?pub_category=Baby%20%26%20Toddler');
    });

    it('preserves original hierarchy format in search URLs', () => {
      const complexCategory = 'Level 1 > Level 2 > Level 3 with spaces';
      renderWithProviders(<CategoryBreadcrumb category={complexCategory} />);

      const level1Link = screen.getByRole('link', { name: 'Level 1' });
      const level2Link = screen.getByRole('link', { name: 'Level 2' });

      expect(level1Link).toHaveAttribute('href', '/s?pub_category=Level%201');
      expect(level2Link).toHaveAttribute('href', '/s?pub_category=Level%201%20%3E%20Level%202');
    });

    it('handles deep hierarchy levels correctly', () => {
      const deepCategory = 'Baby > Feeding > Bottles > Glass > Wide Neck';
      renderWithProviders(<CategoryBreadcrumb category={deepCategory} />);

      const feedingLink = screen.getByRole('link', { name: 'Feeding' });
      const bottlesLink = screen.getByRole('link', { name: 'Bottles' });
      const glassLink = screen.getByRole('link', { name: 'Glass' });

      expect(feedingLink).toHaveAttribute('href', '/s?pub_category=Baby%20%3E%20Feeding');
      expect(bottlesLink).toHaveAttribute('href', '/s?pub_category=Baby%20%3E%20Feeding%20%3E%20Bottles');
      expect(glassLink).toHaveAttribute('href', '/s?pub_category=Baby%20%3E%20Feeding%20%3E%20Bottles%20%3E%20Glass');

      // Last item should not be a link
      expect(screen.getByText('Wide Neck')).toBeInTheDocument();
      expect(screen.queryByRole('link', { name: 'Wide Neck' })).not.toBeInTheDocument();
    });
  });

  describe('translation support', () => {
    it('renders home link with translation key', () => {
      renderWithProviders(<CategoryBreadcrumb category="Baby Products" />);

      // The home link should be present (even if showing translation key)
      const homeLink = screen.getAllByRole('link')[0]; // First link should be home
      expect(homeLink).toHaveAttribute('href', '/s');
    });

    it('maintains correct category links', () => {
      renderWithProviders(<CategoryBreadcrumb category="Products > Clothing > Organic" />);

      const productsLink = screen.getByRole('link', { name: 'Products' });
      const clothingLink = screen.getByRole('link', { name: 'Clothing' });

      expect(productsLink).toHaveAttribute('href', '/s?pub_category=Products');
      expect(clothingLink).toHaveAttribute('href', '/s?pub_category=Products%20%3E%20Clothing');
    });
  });
});