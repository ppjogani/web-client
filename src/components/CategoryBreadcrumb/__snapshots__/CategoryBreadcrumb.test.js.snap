// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryBreadcrumb matches snapshot with hierarchical category 1`] = `
<nav
  aria-label="Category breadcrumb"
  class="root"
>
  <ol
    class="breadcrumbList"
  >
    <li
      class="breadcrumbItem"
    >
      <a
        class="homeLink"
        href="/s"
        title="View all products"
      >
        <span>
          CategoryBreadcrumb.home
        </span>
      </a>
      <span
        aria-hidden="true"
        class="separator"
      >
        &gt;
      </span>
    </li>
    <li
      class="breadcrumbItem"
    >
      <a
        class="categoryLink"
        href="/s?pub_category=Baby%20Products"
        title="View all products in Baby Products"
      >
        Baby Products
      </a>
      <span
        aria-hidden="true"
        class="separator"
      >
        &gt;
      </span>
    </li>
    <li
      class="breadcrumbItem"
    >
      <a
        class="categoryLink"
        href="/s?pub_category=Baby%20Products%20%3E%20Clothing"
        title="View all products in Clothing"
      >
        Clothing
      </a>
      <span
        aria-hidden="true"
        class="separator"
      >
        &gt;
      </span>
    </li>
    <li
      class="breadcrumbItem"
    >
      <span
        aria-current="page"
        class="currentCategory"
      >
        Organic Cotton
      </span>
    </li>
  </ol>
</nav>
`;
